### PR TITLE
Added `make help` target

### PR DESCRIPTION
**Pull request by Bslabe123**
# Description

Running `make help` now lists descriptions for each make target

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/7148

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -86,6 +86,11 @@ ifeq ($(ON_DEFAULT_BRANCH), true)
     ASSETS_ONLY_RELEASE = false
 endif
 
+.PHONY: help
```
Comment (sam-heilbron) [inline Makefile:89]: nit: Could we move this to the top of the file? If I'm a new developer, it's nice to see this first and understand how I can figure out how to inspect this Makefile.

File: Makefile [hunk 2]
```diff
@@ -171,13 +176,13 @@ install-test-tools:
 # requires the environment variable KUBE2E_TESTS to be set to the test type you wish to run
 
 .PHONY: run-tests
-run-tests:
+run-tests: ## Run all tests
```
Comment (sam-heilbron) [inline Makefile:179]: Can we include in the comment some reference to the fact that TEST_PKG env variable can be defined?

File: Makefile [hunk 3]
```diff
@@ -259,9 +263,8 @@ MOCK_RESOURCE_INFO := \
 	gateway:virtual_service:VirtualServiceClient\
 	gateway:route_table:RouteTableClient\
 
-# Use gomock (https://github.com/golang/mock) to generate mocks for our resource clients.
 .PHONY: generate-client-mocks
-generate-client-mocks:
+generate-client-mocks: ## Use gomock (https://github.com/golang/mock) to generate mocks for our resource clients.
```
Comment (sam-heilbron) [inline Makefile:267]: When should a developer run this target? Part of the goal of the help command I think is to figure out which targets are useful for devs.

File: Makefile [hunk 4]
```diff
@@ -786,14 +776,12 @@ SCAN_DIR ?= $(OUTPUT_DIR)/scans
...
-run-security-scan:
-	# Run security scan on gloo and solo-projects
-	# Generates scan files to _output/scans directory
+run-security-scan: ## Run security scan on gloo and solo-projects, Generates scan files to _output/scans directory
 	GO111MODULE=on go run docs/cmd/generate_docs.go run-security-scan -r gloo -a github-issue-latest
 	GO111MODULE=on go run docs/cmd/generate_docs.go run-security-scan -r glooe -a github-issue-latest
 
 .PHONY: publish-security-scan
-publish-security-scan:
+publish-security-scan: ## Publish results generated by run-security-scan
```
Comment (sam-heilbron) [inline Makefile:784]: Should we document this? Is it meant to be run by developers or only in production in CI?

File: changelog/v1.13.0-beta22/ci-stability.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+changelog:
+  - type: NON_USER_FACING
+    description: Add `help` make target to self-document Makefile.
+    issueLink: https://github.com/solo-io/gloo/issues/7148
```
Comment (sam-heilbron) [inline changelog/v1.13.0-beta22/ci-stability.yaml:4]: I don't think this is the appropriate issue link, since this work is independent of leader election changes. I think we can make this non user facing without an issue ref.

File: Makefile [hunk 6]
```diff
@@ -86,6 +86,11 @@ ifeq ($(ON_DEFAULT_BRANCH), true)
     ASSETS_ONLY_RELEASE = false
 endif
 
+.PHONY: help
+help: FIRST_COLUMN_WIDTH=35
```
Comment (sam-heilbron) [inline Makefile:90]: Could we also include a comment to other devs outlining how they can make future targets self-documenting?

**Review by sam-heilbron — COMMENTED** [review]
Looking good! A couple of nits/questions. 

Could you also update the PR to outline what you see when you run `make help` so that the reviewers can confirm that they see the same?

File: Makefile (hunk 1 above)
Comment (sam-heilbron) [reply Makefile:89]: I think we can also make the default goal (https://www.gnu.org/software/make/manual/html_node/Special-Variables.html#Special-Variables) help

File: Makefile [hunk 7]
```diff
@@ -3,6 +3,24 @@
 # run-tests - runs tests (see https://github.com/solo-io/gloo/blob/master/test/e2e/README.md)
 #
 ##########################################################################################
+
```
Comment (jackstine) [inline Makefile:6]: nit: can we get rid of this

Comment (Bslabe123) [reply Makefile:6]: Should we though if were just trying to mirror what we see in solo-projects?

Comment (sam-heilbron) [reply Makefile:6]: I don't think we have to mirror solo-projects identically, ideally we find the best strategy and then mirror that. I think @jackstine is referring to the comment in the first few lines which documents the targets in the file. We can remove that since make help is meant to replace it

File: Makefile [hunk 8]
```diff
@@ -1,8 +1,20 @@
...
+# place a comment after the target that is prefixed by `##`. For example:
+#	custom-target: ## comment that will appear in the documentation when running `make help`
 #
-##########################################################################################
+# **NOTE TO DEVELOPERS**
+# As you encounter make targets that are frequently used, please make them self-documenting
+.PHONY: help
+help: FIRST_COLUMN_WIDTH=35
+help: ## Output the self-documenting make targets
+	@grep -hE '^[%a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-$(FIRST_COLUMN_WIDTH)s\033[0m %s\n", $$1, $$2}'
```
Comment (nfuden) [inline Makefile:15]: this only works on the default make file right? dont we want to have it work on all sub makefiles?

**Review by sam-heilbron — APPROVED** [review]
🎉 awesome addition!

`make help`
```
help                                Output the self-documenting make targets
install-go-tools                    Download and install Go dependencies
run-tests                           Run all tests, or only run the test package at {TEST_PKG} if it is specified
run-ci-regression-tests             Run the Kubernetes E2E Tests in the {KUBE2E_TESTS} package
generate-all                        generated-code
generated-code                      Evaluate go generate
verify-enterprise-protos            Make sure that the enterprise API *.pb.go files that are generated but not used in this repo are valid.
gloo                                Gloo Edge
gloo-docker                         gloo-docker
gloo-race                           Gloo with race detection enabled.
gloo-race-docker                    gloo-race-docker
sds                                 gRPC server for serving Secret Discovery Service config for Gloo Edge MTLS
certgen                             Job for creating TLS Secrets in Kubernetes
build                               Build All
generate-helm-files                 Creates Chart.yaml and values.yaml. See install/helm/README.md for more info.
render-manifests                    Build the Gloo Edge Manifests that are published as release assets
upload-github-release-assets        The code does the proper checking for a {TAGGED_VERSION}
docker-push-retag                   check if all images are already built for {RETAG_IMAGE_REGISTRY}, and build/retag them respectively with tags for the repository specified by {IMAGE_REPO}.
build-test-assets                   Generate Gloo Edge value files and then package the Gloo Edge Helm chart to the _test directory (also generate an index file)
run-security-scan                   Run security scan on gloo and solo-projects, Generates scan files to _output/scans directory
update-licenses                     Check for GPL licenses, if there are any, this will fail
```

`make`
```
help                                Output the self-documenting make targets
install-go-tools                    Download and install Go dependencies
run-tests                           Run all tests, or only run the test package at {TEST_PKG} if it is specified
run-ci-regression-tests             Run the Kubernetes E2E Tests in the {KUBE2E_TESTS} package
generate-all                        generated-code
generated-code                      Evaluate go generate
verify-enterprise-protos            Make sure that the enterprise API *.pb.go files that are generated but not used in this repo are valid.
gloo                                Gloo Edge
gloo-docker                         gloo-docker
gloo-race                           Gloo with race detection enabled.
gloo-race-docker                    gloo-race-docker
sds                                 gRPC server for serving Secret Discovery Service config for Gloo Edge MTLS
certgen                             Job for creating TLS Secrets in Kubernetes
build                               Build All
generate-helm-files                 Creates Chart.yaml and values.yaml. See install/helm/README.md for more info.
render-manifests                    Build the Gloo Edge Manifests that are published as release assets
upload-github-release-assets        The code does the proper checking for a {TAGGED_VERSION}
docker-push-retag                   check if all images are already built for {RETAG_IMAGE_REGISTRY}, and build/retag them respectively with tags for the repository specified by {IMAGE_REPO}.
build-test-assets                   Generate Gloo Edge value files and then package the Gloo Edge Helm chart to the _test directory (also generate an index file)
run-security-scan                   Run security scan on gloo and solo-projects, Generates scan files to _output/scans directory
update-licenses                     Check for GPL licenses, if there are any, this will fail
```

**Review by jackstine — APPROVED** [review]
